fccodegen/_generator.py

```diff
@@ -1,26 +1,180 @@
+
+    import cellmlmanip
+    import cellmlmanip.parser
+
+    model = cellmlmanip.parser.Parser(path).parse()
+    model.make_connections()
+    model.add_units_to_equations()
+    for e in model.equations:
+        model.check_left_right_units_equal(e)
+    return model
+
+
+def get_equations(graph):
+    """
+    Gets the equations from a cellml model graph.
+    """
+    #TODO: This should become part of cellmlmanip
+
+    # Get sorted symbols
+    #TODO: networkx is not a dependency
+    import networkx as nx
+    symbols = nx.lexicographical_topological_sort(graph, key=lambda x: str(x))
+
+    # Create list of equations
+    from sympy.physics.units import Quantity
+    eqs = []
+    for symbol in symbols:
+        # Get equation
+        eq = graph.nodes[symbol]['equation']
+
+        # Skip symbols that are not set with an equation
+        if eq is None:
+            continue
+
+        # Remove quantities (units)
+        eq = eq.subs({q: 1 for q in eq.atoms(Quantity)}, simultaneous=True)
+
+        eqs.append(eq)
+
+    return eqs
+
+
+def get_unique_names(graph):
+    """
+    Creates unique names for all symbols in a cellml model graph.
+    """
+
+    # Create a symbol => name mapping, and a reverse name => symbol mapping
+    symbols = {}
+    reverse = {}
+
+    def uname(name):
+        """ Add an increasing number to a name until it's unique """
+        root = name + '_'
+        i = 0
+        while name in reverse:
+            i += 1
+            name = root + str(i)
+        return name
+
+    for v in graph:
+        if isinstance(v, sp.Derivative):
+            continue
+
+        # Try simple name
+        parts = v.name.split('$')
+        assert len(parts) == 2
+        name = parts[-1]
+
+        # If already taken, rename _both_ variables using component name
+        if name in reverse:
+
+            # Get existing variable
+            other = reverse[name]
+
+            # Check it hasn't been renamed already
+            if symbols[other] == name:
+                oparts = other.name.split('$')
+                assert len(oparts) == 2
+                oname = uname(oparts[0] + '_' + oparts[1])
+                symbols[other] = oname
+                reverse[oname] = other
+
+            # Get new name for v
+            name = uname(parts[0] + '_' + parts[1])
+
+        # Store symbol name
+        symbols[v] = name
+        reverse[name] = v
+
+    return symbols
+
+
+def get_state_list(graph):
+    """
+    Returns a list of state variables found in the given model graph.
+    """
+    states = []
+    for v in graph:
+        if isinstance(v, sp.Derivative):
+            states.append(v.args[0])
+    return states
+
+
 def create_weblab_model(model, path):
     """
     Takes a :class:`cellmlmanip.Model`, generates a ``.pyx`` model for use
     with the Web Lab, and stores it at ``path``.
     """
+    # Get equation graph
+    graph = model.get_equation_graph()
+
+    # Get ordered equations
```

Comment:
Eventually when more API methods are in `cellmlmanip` this needs to happen below the definitions of `symbol_name` etc. and look more like:
```python
state_vars = model.get_state_variables()
free_var = model.get_free_variable()
derivatives = [sp.Derivative(state, free_var) for state in state_vars]
eqs = model.get_equations_for(derivatives)
```

Reply:
Agreed!